lib/Parse/ParseExpr.cpp

```diff
@@ -3122,19 +3138,95 @@ ParserStatus Parser::parseExprList(tok leftTok, tok rightTok,
     return status;
 
   // Parse the closure.
-  ParserResult<Expr> closure =
-    parseTrailingClosure(SourceRange(leftLoc, rightLoc));
-  status |= closure;
-  if (closure.isNull())
-    return status;
+  status |=
+      parseTrailingClosures(isExprBasic, SourceRange(leftLoc, rightLoc),
+                            trailingLBrace, trailingRBrace, trailingClosures);
+  return status;
+}
 
-  // Record the trailing closure.
-  trailingClosure = closure.get();
+/// Check whether upcoming trailing closure actually represents a block of
+/// labeled trailing closures e.g.
+/// foo() {
+///  <label-1>: { ... }
+///  ...
+/// }
+static bool isBlockOfMultipleTrailingClosures(bool isExprBasic, Parser &P) {
+  assert(isValidTrailingClosure(isExprBasic, P));
+
+  // If closure doesn't start from a label there is no reason
+  // to look any farther.
+  {
+    auto nextTok = P.peekToken();
+    if (!nextTok.canBeArgumentLabel() || nextTok.is(tok::kw__) ||
+        nextTok.getText()[0] == '$')
+      return false;
+  }
 
-  return status;
+  Parser::BacktrackingScope backtrack(P);
+  P.consumeToken(tok::l_brace);
+  P.consumeToken(); // Consume Label text.
+
+  if (!P.Tok.is(tok::colon))
+    return false;
+
+  P.consumeToken(); // Consume `:`
+
+  // If the next token after the label is not a left brace
+  // that means what follows is not a valid closure, which
+  // in turn means it can't be a labeled block.
+  if (!P.Tok.is(tok::l_brace))
+    return false;
+
+  return isValidTrailingClosure(isExprBasic, P);
+}
+
+ParserStatus Parser::parseMultipleTrailingClosures(
+    SourceLoc &LBrace, SourceLoc &RBrace,
+    SmallVectorImpl<TrailingClosure> &closures) {
+  // Consume '{' of the trailing closure
+  LBrace = consumeToken(tok::l_brace);
+
+  // There could N labeled closures depending on the number of arguments.
+  do {
+    if (!(Tok.canBeArgumentLabel() && peekToken().is(tok::colon))) {
+      diagnose(Tok.getLoc(),
+               diag::expected_argument_label_followed_by_closure_literal);
+      skipUntilDeclStmtRBrace(tok::r_brace);
+      break;
+    }
+
+    Identifier label;
+    auto labelLoc = consumeArgumentLabel(label);
+    if (!labelLoc.isValid())
+      return makeParserError();
+
+    consumeToken(); // Consume ':'
+
+    auto closure = parseExprClosure();
+    if (closure.isNull())
+      return makeParserError();
+
+    closures.push_back({label, labelLoc, closure.get()});
+
+    // Recover if blocks are separated by comma instead of newline.
+    if (Tok.is(tok::comma)) {
+      diagnose(Tok.getLoc(), diag::unexpected_separator, ",")
+        .fixItRemove(Tok.getLoc());
+      consumeToken();
+    }
+  } while (!Tok.is(tok::r_brace));
+
+  // Consume `}` of the trailing closure.
+  parseMatchingToken(tok::r_brace, RBrace,
+                     diag::expected_multiple_closures_block_rbrace, LBrace);
+
+  return makeParserSuccess();
```

Comment:
Could you propagate the status from `parseExprClosure()`s? Like:
```cpp
ParserStatus Status;
do {
   ...
   auto closure = parseExprClosure();
   Status |= closure;
   ...
} while (...);

if (parseMatchingToken(...)) {
  Status.setIsParseError()
}

return Status;
```
Because we want to know if a code completion happens.
You don't need to care about code completion at the labels position, because we can do it later.